internal/api/installation_db_restoration_operation.go

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+//
+
+package api
+
+import (
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/mattermost/mattermost-cloud/internal/common"
+	"github.com/mattermost/mattermost-cloud/model"
+)
+
+// initInstallationRestoration registers installation restoration operation endpoints on the given router.
+func initInstallationRestoration(apiRouter *mux.Router, context *Context) {
+	addContext := func(handler contextHandlerFunc) *contextHandler {
+		return newContextHandler(context, handler)
+	}
+
+	restorationsRouter := apiRouter.PathPrefix("/operations/database/restorations").Subrouter()
+
+	restorationsRouter.Handle("", addContext(handleTriggerInstallationDBRestoration)).Methods("POST")
+	restorationsRouter.Handle("", addContext(handleGetInstallationDBRestorationOperations)).Methods("GET")
+
+	restorationRouter := apiRouter.PathPrefix("/operations/database/restoration/{restoration:[A-Za-z0-9]{26}}").Subrouter()
+	restorationRouter.Handle("", addContext(handleGetInstallationDBRestorationOperation)).Methods("GET")
+}
+
+// handleTriggerInstallationDBRestoration responds to POST /api/installations/operations/database/restorations,
+// requests restoration of Installation's data.
+func handleTriggerInstallationDBRestoration(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.Logger = c.Logger.
+		WithField("action", "restore-installation-database")
+
+	restoreRequest, err := model.NewInstallationDBRestorationRequestFromReader(r.Body)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to decode request")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	c.Logger = c.Logger.
+		WithField("installation", restoreRequest.InstallationID).
+		WithField("backup", restoreRequest.BackupID)
+
+	newState := model.InstallationStateDBRestorationInProgress
+
+	installationDTO, status, unlockOnce := getInstallationForTransition(c, restoreRequest.InstallationID, newState)
+	if status != 0 {
+		w.WriteHeader(status)
+		return
+	}
+	defer unlockOnce()
+
+	backup, err := c.Store.GetInstallationBackup(restoreRequest.BackupID)
+	if err != nil {
+		c.Logger.WithError(err).Errorf("failed to get backup")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if backup == nil {
+		c.Logger.Error("Backup not found")
+		w.WriteHeader(http.StatusBadRequest)
```

Comment:
`http.StatusNotFound` might be better?

Reply:
Yea, I have chosen to use `400` as this ID is passed in request body not in the path. I am 0/5 on that tho, if you feel strong about it I can change it.

Reply:
Yeah, 1/5 on 404 making a bit more sense to me, but good with either.